crates/precompile/src/crypto_provider/mod.rs

```diff
@@ -0,0 +1,567 @@
+    ///
+    /// # Returns
+    /// The sum as 192 bytes (unpadded), or an error if points are invalid.
+    fn bls12_381_g2_add(
+        &self,
+        a: bls12_381::G2Point,
+        b: bls12_381::G2Point,
+    ) -> Result<[u8; 192], PrecompileError>;
+
+    /// BLS12-381 G1 multi-scalar multiplication.
+    ///
+    /// # Arguments
+    /// * `points_scalars` - Vector of (point, scalar) pairs where:
+    ///   - Points are G1Point tuples: (x, y) with each component being 48 bytes
+    ///   - Scalars are 32 bytes each
+    ///
+    /// # Returns
+    /// The result as 96 bytes (unpadded), or an error if inputs are invalid.
+    fn bls12_381_g1_msm(
+        &self,
+        points_scalars: &[(bls12_381::G1Point, [u8; 32])],
+    ) -> Result<[u8; 96], PrecompileError>;
+
+    /// BLS12-381 G2 multi-scalar multiplication.
+    ///
+    /// # Arguments
+    /// * `points_scalars` - Vector of (point, scalar) pairs where:
+    ///   - Points are G2Point tuples: (x0, x1, y0, y1) with each component being 48 bytes
+    ///   - Scalars are 32 bytes each
+    ///
+    /// # Returns
+    /// The result as 192 bytes (unpadded), or an error if inputs are invalid.
+    fn bls12_381_g2_msm(
+        &self,
+        points_scalars: &[(bls12_381::G2Point, [u8; 32])],
+    ) -> Result<[u8; 192], PrecompileError>;
+
+    /// BLS12-381 pairing check.
+    ///
+    /// # Arguments
+    /// * `pairs` - Vector of PairingPair (G1, G2) point pairs for pairing check
+    ///
+    /// # Returns
+    /// `true` if the pairing check passes, `false` otherwise.
+    fn bls12_381_pairing(&self, pairs: &[bls12_381::PairingPair]) -> Result<bool, PrecompileError>;
+
+    /// BLS12-381 map field element to G1.
+    ///
+    /// # Arguments
+    /// * `fp` - Field element as 48 bytes
+    ///
+    /// # Returns
+    /// The mapped G1 point as 96 bytes (unpadded).
+    fn bls12_381_map_fp_to_g1(&self, fp: &[u8; 48]) -> Result<[u8; 96], PrecompileError>;
+
+    /// BLS12-381 map field element to G2.
+    ///
+    /// # Arguments
+    /// * `fp2` - Fp2 element as two 48-byte field elements
+    ///
+    /// # Returns
+    /// The mapped G2 point as 192 bytes (unpadded).
+    fn bls12_381_map_fp2_to_g2(
+        &self,
+        fp2_x: &[u8; 48],
+        fp2_y: &[u8; 48],
+    ) -> Result<[u8; 192], PrecompileError>;
+
+    /// KZG point evaluation.
+    ///
+    /// Verifies a KZG proof for polynomial evaluation.
+    ///
+    /// # Arguments
+    /// * `commitment` - The KZG commitment (48 bytes)
+    /// * `z` - The evaluation point (32 bytes, big-endian)
+    /// * `y` - The claimed evaluation result (32 bytes, big-endian)
+    /// * `proof` - The KZG proof (48 bytes)
+    ///
+    /// # Returns
+    /// `true` if the proof is valid, `false` otherwise.
+    #[cfg(any(feature = "c-kzg", feature = "kzg-rs"))]
+    fn kzg_verify_proof(
+        &self,
+        commitment: &[u8; 48],
+        z: &[u8; 32],
+        y: &[u8; 32],
+        proof: &[u8; 48],
+    ) -> bool;
+
```

Comment:
Preserving the old behaviour, essentially if c-kzg or kzg-rs are not enabled, then this is not available. It should be possible to reuse one of the bls12_381 crates to implement this if needed